rust/protocol/src/error.rs

```diff
@@ -11,82 +15,208 @@ use uuid::Uuid;
     UnrecognizedMessageVersion(u32),
 
     /// fingerprint version number mismatch them {0} us {1}
+    ///
+    /// Raised if a fingerprint version decoded from a protobuf has an unexpected value.
     FingerprintVersionMismatch(u32, u32),
     /// fingerprint parsing error
+    ///
+    /// Raised if a field in a fingerprint protobuf is invalid in some way.
+    ///
+    /// Similar to [Self::InvalidProtobufEncoding].
     FingerprintParsingError,
 
     /// no key type identifier
+    ///
+    /// Raised if a [PublicKey] is deserialized from an empty slice.
+    ///
+    /// Prefer to use static-sized slices in API method signatures and struct fields to minimize the
+    /// need to raise this error.
     NoKeyTypeIdentifier,
     /// bad key type <{0:#04x}>
+    ///
+    /// Raised if a [KeyType] decoded from a [u8] has an unrecognized value.
     BadKeyType(u8),
     /// bad key length <{1}> for key with type <{0}>
+    ///
+    /// Raised if a [PublicKey] or [PrivateKey] is deserialized from a slice of incorrect length.
+    ///
+    /// Prefer to use static-sized slices in API method signatures and struct fields to minimize the
+    /// need to raise this error.
     BadKeyLength(KeyType, usize),
 
     /// invalid signature detected
+    ///
+    /// Raised if signature validation fails for a [SignedPreKeyRecord] or a [SenderKeyMessage].
     SignatureValidationFailed,
 
     /// untrusted identity for address {0}
+    ///
+    /// Raised if an identity verification check fails in [message_encrypt] or [message_decrypt].
     UntrustedIdentity(crate::ProtocolAddress),
 
     /// invalid prekey identifier
+    ///
+    /// Raised if a [PreKeyId] could not be resolved to a [PreKeyRecord] by a [PreKeyStore].
     InvalidPreKeyId,
     /// invalid signed prekey identifier
+    ///
+    /// Raised if a [SignedPreKeyId] could not be resolved to a [SignedPreKeyRecord] by
+    /// a [SignedPreKeyStore].
     InvalidSignedPreKeyId,
 
     /// invalid MAC key length <{0}>
+    ///
+    /// Raised if a MAC key is deserialized from an incorrectly-sized slice.
+    ///
+    /// Prefer to use static-sized slices in API method signatures and struct fields to minimize the
+    /// need to raise this error.
     InvalidMacKeyLength(usize),
 
     /// missing sender key state for distribution ID {distribution_id}
-    NoSenderKeyState { distribution_id: Uuid },
+    ///
+    /// Raised if a [SenderKeyStore] is unable to locate a [SenderKeyRecord] for a given
+    /// *([ProtocolAddress], [Uuid])* pair in [group_encrypt] or [group_decrypt].
+    NoSenderKeyState {
+        /// Unique identifier for the sender key session.
+        distribution_id: Uuid,
+    },
 
     /// session with {0} not found
+    ///
+    /// Raised if an [IdentityKeyStore] does not contain an entry for a [ProtocolAddress], or
+    /// alternately if a [SessionStore] does not contain a session for a given [ProtocolAddress].
     SessionNotFound(crate::ProtocolAddress),
```

Comment:
I'd switch the order of these; the *common* use is a missing session (hence the name), but it also gets used for missing identity when one is expected because repairing the session will fetch the identity too. cf. comment in the middle of `sealed_sender_multi_recipient_encrypt`:

```
// Returned as a SessionNotFound error because (a) we don't have an identity error
// that includes the address, and (b) re-establishing the session should re-fetch
// the identity.
```